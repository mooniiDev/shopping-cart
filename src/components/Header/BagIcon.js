// Font Awesome imports
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBagShopping } from '@fortawesome/free-solid-svg-icons';

function BagIcon() {
  return (
    <button type="button" className="self-center text-3xl">
      <FontAwesomeIcon
        icon={faBagShopping}
        className="fa-fw"
        pointerEvents="none"
      />
    </button>
  );
}

export default BagIcon;
