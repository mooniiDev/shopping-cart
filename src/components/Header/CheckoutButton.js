// Font Awesome imports
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faShoppingBag } from '@fortawesome/free-solid-svg-icons';

function CheckoutButton() {
  return (
    <button type="button" className="self-center text-3xl">
      <FontAwesomeIcon
        icon={faShoppingBag}
        className="fa-fw"
        pointerEvents="none"
      />
    </button>
  );
}

export default CheckoutButton;
