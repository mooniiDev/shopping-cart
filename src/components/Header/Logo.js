function Logo() {
  return (
    <h1>
      <a href="/" className="font-rampart">
        PIXEDELIC
      </a>
    </h1>
  );
}

export default Logo;
