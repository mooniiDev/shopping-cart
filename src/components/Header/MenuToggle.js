// Font Awesome imports
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars } from '@fortawesome/free-solid-svg-icons';

function MenuToggle() {
  return (
    <div className="flex items-center">
      <button type="button" className="self-center text-3xl">
        <FontAwesomeIcon icon={faBars} className="fa-fw" pointerEvents="none" />
      </button>
    </div>
  );
}

export default MenuToggle;
