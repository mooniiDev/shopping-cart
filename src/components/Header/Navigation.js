function Navigation() {
  return (
    <nav>
      <ul className="flex text-xl md:gap-10">
        <li className="flex h-16 w-1/2 justify-center bg-persian-pink md:h-11 md:bg-opacity-0">
          <a href="/" className="self-center font-bold">
            HOME
          </a>
        </li>
        <li className="flex h-16 w-1/2 justify-center bg-light-green md:h-11 md:bg-opacity-0">
          <a href="/shop" className="self-center font-bold">
            SHOP
          </a>
        </li>
      </ul>
    </nav>
  );
}

export default Navigation;
