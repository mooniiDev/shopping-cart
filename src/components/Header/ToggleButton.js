// Font Awesome imports
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars } from '@fortawesome/free-solid-svg-icons';

function ToggleButton() {
  return (
    <button
      type="button"
      className="self-center text-3xl"
      data-testid="toggle-button"
    >
      <FontAwesomeIcon icon={faBars} className="fa-fw" pointerEvents="none" />
    </button>
  );
}

export default ToggleButton;
