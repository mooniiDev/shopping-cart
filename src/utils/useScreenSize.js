import { useState, useEffect } from 'react';

// A custom hook that allows to monitor changes in the screen size
// and return a boolean value that can be used to conditionally
// render components based on the screen size
function useScreenSize(query) {
  const [isMobile, setIsMobile] = useState(false);

  // A callback function, that updates the `isMobile` state variable
  // whenever the screen size changes
  function handleScreenSizeChange(event) {
    setIsMobile(event.matches);
  }

  useEffect(() => {
    // Check if `window` object exists and if
    // the `window.matchMedia` method is available before using it
    if (typeof window !== 'undefined' && window.matchMedia) {
      // Define a `MediaQueryList` object using the `window.matchMedia` method
      // Pass in a media query string that checks if the screen width
      // is less than or equal to the given size
      const mediaQuery = window.matchMedia(query);
      // Set the initial value of the `isMobile` state variable
      // to the result of the `matches` property of the `MediaQueryList` object
      setIsMobile(mediaQuery.matches);
      // Add an event listener to the `MediaQueryList` object that listens
      // for changes in the screen size and call `handleScreenSizeChange`
      // function when a change occurs
      mediaQuery.addEventListener('change', handleScreenSizeChange);

      return () => {
        mediaQuery.removeEventListener('change', handleScreenSizeChange);
      };
    }
    // In cases where the `window` object is not defined or the `window.matchMedia` method is not available,
    // we cannot add the event listener and hence, we do not need a cleanup function.
    // In such cases, we return an empty arrow function `() => {}` from the `useEffect` hook
    // to satisfy the `return` requirement and avoid any potential errors.
    return () => {};
  }, [query]);

  return isMobile;
}

export default useScreenSize;
