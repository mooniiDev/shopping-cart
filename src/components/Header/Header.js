// Custom hook of monitoring screen size changes
import useScreenSize from '../../utils/useScreenSize';

// Components imports
import Logo from './Logo';
import ToggleButton from './ToggleButton';
import Navigation from './Navigation';
import CheckoutButton from './CheckoutButton';

function Header() {
  const isMobile = useScreenSize('(max-width: 767px)');

  return (
    <header className="bg-tropical-indigo text-white md:p-6">
      {isMobile ? (
        <div className="flex flex-col">
          <div className="flex justify-between p-6">
            <ToggleButton />
            <Logo />
            <CheckoutButton />
          </div>
          <Navigation />
        </div>
      ) : (
        <div className="flex justify-around">
          <Logo />
          <Navigation />
          <CheckoutButton />
        </div>
      )}
    </header>
  );
}

export default Header;
