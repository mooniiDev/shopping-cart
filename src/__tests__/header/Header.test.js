import React from 'react';
import { render } from '@testing-library/react';
import Header from '../../components/Header/Header';
import useScreenSize from '../../utils/useScreenSize';

jest.mock('../../utils/useScreenSize');

describe('Header', () => {
  describe('on mobile', () => {
    beforeEach(() => {
      useScreenSize.mockImplementation(() => true);
    });

    it('renders correctly', () => {
      const tree = render(<Header />).container;
      expect(tree).toMatchSnapshot('mobile');
    });
  });

  describe('on desktop', () => {
    beforeEach(() => {
      useScreenSize.mockImplementation(() => false);
    });

    it('renders correctly', () => {
      const tree = render(<Header />).container;
      expect(tree).toMatchSnapshot('desktop');
    });
  });
});
