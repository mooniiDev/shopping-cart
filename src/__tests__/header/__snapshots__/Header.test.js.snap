// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header on desktop renders correctly: desktop 1`] = `
<div>
  <header
    class="bg-tropical-indigo text-white md:p-6"
    data-testid="header"
  >
    <div
      class="flex justify-around"
    >
      <h1
        class="mb-1.5 flex md:w-52"
        data-testid="logo"
      >
        <a
          class="font-rampart md:self-center"
          href="/"
        >
          PIXEDELIC
        </a>
      </h1>
      <nav
        data-testid="navigation"
      >
        <ul
          class="flex text-xl md:gap-10"
        >
          <li
            class="flex h-16 w-1/2 justify-center bg-persian-pink md:h-11 md:bg-opacity-0"
          >
            <a
              class="self-center font-bold"
              href="/"
            >
              HOME
            </a>
          </li>
          <li
            class="flex h-16 w-1/2 justify-center bg-light-green md:h-11 md:bg-opacity-0"
          >
            <a
              class="self-center font-bold"
              href="/shop"
            >
              SHOP
            </a>
          </li>
        </ul>
      </nav>
      <button
        class="flex justify-end text-3xl md:w-52"
        data-testid="checkout-button"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bag-shopping fa-fw self-center"
          data-icon="bag-shopping"
          data-prefix="fas"
          focusable="false"
          pointer-events="none"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M160 112c0-35.3 28.7-64 64-64s64 28.7 64 64v48H160V112zm-48 48H48c-26.5 0-48 21.5-48 48V416c0 53 43 96 96 96H352c53 0 96-43 96-96V208c0-26.5-21.5-48-48-48H336V112C336 50.1 285.9 0 224 0S112 50.1 112 112v48zm24 48a24 24 0 1 1 0 48 24 24 0 1 1 0-48zm152 24a24 24 0 1 1 48 0 24 24 0 1 1 -48 0z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </header>
</div>
`;

exports[`Header on mobile renders correctly: mobile 1`] = `
<div>
  <header
    class="bg-tropical-indigo text-white md:p-6"
    data-testid="header"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-between p-6"
      >
        <button
          class="self-center text-3xl"
          data-testid="toggle-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bars fa-fw"
            data-icon="bars"
            data-prefix="fas"
            focusable="false"
            pointer-events="none"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
              fill="currentColor"
            />
          </svg>
        </button>
        <h1
          class="mb-1.5 flex md:w-52"
          data-testid="logo"
        >
          <a
            class="font-rampart md:self-center"
            href="/"
          >
            PIXEDELIC
          </a>
        </h1>
        <button
          class="flex justify-end text-3xl md:w-52"
          data-testid="checkout-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bag-shopping fa-fw self-center"
            data-icon="bag-shopping"
            data-prefix="fas"
            focusable="false"
            pointer-events="none"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M160 112c0-35.3 28.7-64 64-64s64 28.7 64 64v48H160V112zm-48 48H48c-26.5 0-48 21.5-48 48V416c0 53 43 96 96 96H352c53 0 96-43 96-96V208c0-26.5-21.5-48-48-48H336V112C336 50.1 285.9 0 224 0S112 50.1 112 112v48zm24 48a24 24 0 1 1 0 48 24 24 0 1 1 0-48zm152 24a24 24 0 1 1 48 0 24 24 0 1 1 -48 0z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <nav
        data-testid="navigation"
      >
        <ul
          class="flex text-xl md:gap-10"
        >
          <li
            class="flex h-16 w-1/2 justify-center bg-persian-pink md:h-11 md:bg-opacity-0"
          >
            <a
              class="self-center font-bold"
              href="/"
            >
              HOME
            </a>
          </li>
          <li
            class="flex h-16 w-1/2 justify-center bg-light-green md:h-11 md:bg-opacity-0"
          >
            <a
              class="self-center font-bold"
              href="/shop"
            >
              SHOP
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>
</div>
`;
